feat(AccessControl): Add how to protect BDM instances

danila-m opened the pull request:
Relates to :[BS-17657](https://bonitasoft.atlassian.net/browse/BS-17657)

danila-m committed:
feat(AccessControl): Add how to protect BDM instances

Relates to :[BS-17657](https://bonitasoft.atlassian.net/browse/BS-17657)

educhastenier commented:
test this please

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+        long currentUserId = session.getUserId()
+        List<Profile> profilesForUser = apiAccessor.getProfileAPI().getProfilesForUser(currentUserId, 0, 50, ProfileCriterion.ID_ASC)
+        if (!apiCallContext.getQueryString().contains("q=findBySubject")) {
+            println "q=findBySubject found in query string"
```
remove this line

educhastenier commented on md/rest-api-authorization.md:
```diff
@@ -193,7 +193,7 @@ import org.json.JSONObject
 
     @Override
-    public boolean check(APISession apiSession, APICallContext apiCallContext, APIAccessor apiAccessor, Logger logger) {
+    boolean isAllowed(APISession apiSession, APICallContext apiCallContext, APIAccessor apiAccessor, Logger logger) {
```
good catch

danila-m committed:
title formatting

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+
+## III - Instances protection
+
+Instance protection is not supported as a part of the access control feature. However it is still possible to solve most use cases using the [rest-api authorizations](rest-api-authorization.md).
```
as a part of the BDM Access Control

educhastenier replied:
to be more precise

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+
+::: info
+The method shown in the example below is not a new 7.7 feature, it is present in the Community version since 7.0 ( although the location of the files and the push/pull procedure changed in 7.6.0).
+It allows to protect the bdm instances by limiting the access to the jpql requests that retrieve the object instances rather than to the instances themselves.
```
~jpql requests~ => BDM query requests

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+**Scenario:**
+
+We will look into a mark contestation process for a school. Students can make requests about their marks to their teachers through a BPM process.
+Each teacher teaches a different subject. A teacher can only access requests made concerning his subject.
```
~his~ => their

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+
+**1. Define the BDM**
+
+We will use a simple relatively simple BDM for this use case: 
```
~simple~ relatively simple

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+
+We will use a simple relatively simple BDM for this use case: 
+
+1. an object Student
```
an object Student with attributes:

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+|---|---|
+| fullname | STRING |
+
+2. an object Request
```
...with attributes:

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+
+**2. Define the profiles**
+
+We will need a few new profiles, mapped to one or more users. We need one profile per subject (), as well as a **Teacher** profile (for all the teachers). A maths teacher would therefore have **MathematicsTeacher** and a **Teacher** profile. In the rest of the guide we will use 3 profiles **MathematicsTeacher**, **PhysicsTeacher** and **Teacher**, mapped to **helen.kelly**,**jan.fischer** and bot ...
```
() ?.

educhastenier replied:
**helen.kelly**, **jan.fischer** (missing space char)

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+
+**3. Create a process and generate some request instances**
+
+You will need some instances of the request object, as well as some students. To create them, follow the steps discribed in the section  **II - Invoice**, part 3 *Create a process and generate some data* of this very page.
```
Add direct link to the §

educhastenier replied:
link does not work

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+- Add  a new Table widget to the page (To display the requests):
+    - **Headers** : Id, Subject, Content, Medical comment, Student (constants)
+    - **Content** : requestList (script)
+    - **Column keys** : persistenceId, subject, content, medicalComment, student.fullname
```
Those 2 blocks do not show well separated. We think it is the same variable

educhastenier replied:
still bad

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+    
+- Add  a new Table widget to the page (To display the requests):
+    - **Headers** : Id, Subject, Content, Medical comment, Student (constants)
+    - **Content** : requestList (script)
```
**Content** : requestList (script, click the X icon to switch from contstant to script (bound data?))

educhastenier replied:
~X~ => **fx**

educhastenier commented on md/bdm-access-control.md:
````diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+- Go to */BonitaStudioSubscription-7.7.0/workspace/tomcat/setup/platform_conf/current/tenants/1/tenant_security_scripts*. Create a file *SubjectTeacherPermissionRule.groovy*, with the following content:
+
+```groovy
+package org.bonitasoft.permissions
````
remove the package declaration. we did it on your IntelliJ installation, but in real life, there must be NO package (default unnamed package)

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+        APISession session = apiSession
+        long currentUserId = session.getUserId()
+        List<Profile> profilesForUser = apiAccessor.getProfileAPI().getProfilesForUser(currentUserId, 0, 50, ProfileCriterion.ID_ASC)
+        if (!apiCallContext.getQueryString().contains("q=findBySubject")) {
```
I would add comments to indicate what we do. Namely here:
```
# First, let's check we only restrict access to query named "findBySubject":
```

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+        if (filters.containsKey("subject")) {
+            def subjectAsString = filters.get("subject")
+            for (Profile p : profilesForUser) {
+                if (p.name.contains(subjectAsString)){
```
```
# Let's check the logged-in user (teacher) has at least one profile matching the Class Subject:
```

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+            }
+            return false
+        }
+        return true
```
```
# otherwise, it's an access to a different query, so no filtering in this case:

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+- Return to *BonitaStudioSubscription-7.7.0/workspace/tomcat/setup/* and run *setup push*. this will upload the *dynamic-permissions-checks-custom.properties* file to the server.
+- Restart your web server. The new security rule from the *dynamic-permissions-checks-custom.properties* file now active.
+::: warning
+	**NB:** For every modification of the *dynamic-permissions-checks-custom.properties* file, you will need to push it and restart the web server. However, since the Studio has the Debug mode active by default, you don't need to restart the web server after modifying the groovy script in this environnement.
```
Remove tabs

educhastenier commented on md/bdm-access-control.md:
````diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+```
+
+- Return to *BonitaStudioSubscription-7.7.0/workspace/tomcat/setup/* and run *setup push*. this will upload the *dynamic-permissions-checks-custom.properties* file to the server.
+- Restart your web server. The new security rule from the *dynamic-permissions-checks-custom.properties* file now active.
````
Restart your web server through menu "Server -> Restart Web server" (or whatever is the real name)...
...**is** now active

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +314,153 @@ Deploy your access control file.
+**6. Adding access control on attributes**
+
+It is possible to use both this security and the accessControl feature. For example, if you decide that the field Medical Comment should not be visible to teachers, you can upload an accessControl file, and the field won't be returned by the *findBySubject* request.
+
```
See above for attribute filtering.

educhastenier commented on md/rest-api-authorization.md:
```diff
@@ -304,6 +304,12 @@ For example, to add the permission to the user walter.bates (username), add the
 
+## Authorizing access to a BDM object or its attributes
+
+Starting with the Bonita efficiency subscription version, you can use a simpler mechanism to grant or deny access to BDM objects or some of their attributes to specific profiles, using the BDM accessControl feature.
```
~version~ -> edition

educhastenier replied:
using the BDM Access Control feature

educhastenier commented on md/rest-api-authorization.md:
```diff
@@ -304,6 +304,12 @@ For example, to add the permission to the user walter.bates (username), add the
 
 <a id="activate"/>
 
+## Authorizing access to a BDM object or its attributes
```
~Authorizing~ -> Restricting

danila-m committed:
manu feedback integration

danila-m committed:
manu feedback integration 2

alachambre commented:
It would be nice to add a small sentence in the overview (at the begging) . At the moment, it is written that we can only protect attributes, thus there is an instance protection part, it's a bit confusing.

danila-m committed:
adrien feedback integration

danila-m commented:
Following Adrien's review, I made a proposition concerning the overview part of the page.

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -5,8 +5,7 @@
 ## Overview
-Currently, in Bonita, defining access control on a business object (BO) means defining which attributes of this business object a given user is allowed to access through REST API.  
-If access controls are defined on the BDM, when a user calls a BDM API to retrieve some objects, the response will only contain the attributes this user is allowed to see.
+This page explains how to protect the access to your BDM. It explores all the methods available in Bonita, both legacy methods (through [Rest api authorizations](rest-api-authorization.md)), and new 7.7 Access Control feature.
```
Not sure it is clearer this way, but I let others decide...

alachambre replied:
agree, i would do something like

-Currently, in Bonita, defining access control on a business object (BO) means defining which attributes of this business object a given user is allowed to access through REST API.  The access control feature doesn't give the possibility to define which instance of a business object a user is allowed to access. It is still possible to do so using Bonita, see exemple 3.  
 -If access controls are defined on the BDM, when a user calls a BDM API to retrieve some objects, the response will only contain the attributes this user is allowed to see.

danila-m replied:
I will let nath decide on this one.

NathalieC replied:
This page explains how to protect the access to your BDM through REST API calls. 
The BDM access control feature, available in Bonita Studio, allows to grant access to entire business objects or access to some attributes of business objects.
To grant access to specific instances of business objects, you will need to use [Rest api authorizations](rest-api-authorization.md).
You will find below examples of both use-cases.

alachambre commented on md/bdm-access-control.md:
```diff
@@ -314,3 +315,157 @@ Deploy your access control file.
+| subject | STRING |
+| medicalComment | STRING |
+| content | STRING |
+| student | Student |
```
I think you should specify the relation type of the student, it makes a huge difference in the way to generate data.

educhastenier replied:
true

danila-m replied:
Done.

alachambre commented on md/bdm-access-control.md:
```diff
@@ -314,3 +315,157 @@ Deploy your access control file.
+**3. Create a process and generate some request instances**
+
+You will need some instances of the request object, as well as some students. To create them, follow the steps discribed in the section  [II - Invoice](#bdmFilling).
+
```
Tu user must create data with Mathematics and Physics as subjects, else the application won't work if he follows the next step, so it should be specified here .

alachambre commented on md/bdm-access-control.md:
```diff
@@ -314,3 +315,157 @@ Deploy your access control file.
+Now in the Studio, create a new [application](applicationCreation.md). Call it *TeacherApp*, give it a theme, and a homepage token, and map it to the *Teacher* profile.
+Then create a one-page menu *Student Request*, with your homepage token and add to it the page you just created in the UID.
+
+If you preview the page at this point, both jan.fischer and helen.kelly can access your new application, select from a drop down list either *Mathematics* or *Physics*, wich should display all the Mathematics and all the Physics requests, respectively, for both users.
```
which

alachambre commented on md/bdm-access-control.md:
```diff
@@ -314,3 +315,157 @@ Deploy your access control file.
+**5. Security implementation**
+
+- Got to *BonitaStudioSubscription-7.7.0/workspace/tomcat/setup/*
+- Run *setup pull*
```
* Something is missing, the user should set the path to the H2 directory before to run setup pull
* It would be nice to add the link to this page, just in case the user wonders what's this setup pull stuff and what does it do: https://documentation.bonitasoft.com/bonita/7.6/BonitaBPM_platform_setup
* There is a well detailed example on how to configure the dynamic security in the rest api authorization documentation, especially about all the stuff in the platform conf folder. it could be interesting to copy / link it here , to help the user to understand what's going on. https://documentation.bonitasoft.com/bonita/7.6/rest-api-authorization

alachambre commented on md/bdm-access-control.md:
```diff
@@ -314,3 +315,157 @@ Deploy your access control file.
+        def filters = apiCallContext.getFilters()
+        if (filters.containsKey("subject")) {
+            def subjectAsString = filters.get("subject")
+            //Let's check the logged-in user (teacher) has at least one profile matching the Class Subject:
```
maybe add a small example on the comment to help the user to understand the logic: 
subjectAsString == Physics 
profile PhysicsTeacher contains physics -> ok
profile MathematicsTeacher doesn't contain Physics -> not ok
=> only someone with the profile PhysicsTeacher will have th authorization

alachambre reviewed (changes requested):
some stuff to change in my view, especially in the setup pull / push part

danila-m committed:
adrien feedback integration 2

educhastenier commented on md/bdm-access-control.md:
````diff
@@ -405,16 +405,18 @@ In the UI Designer, create an application page (*reviewRequests*):
 - Open the *dynamic-permissions-checks-custom.properties*, and add the following line:
 ``` GET|bdm/businessData/com.company.model.Request=[check|SubjectTeacherPermissionRule] ```
-- This line indicates that before executing any queries on the com.company.model.Request object types in the BDM, a verification has to be run. In this case we indicate that the verification is a groovy script, *SubjectTeacherPermissionRule.groovy*, which we will now create.
+- This line indicates that before executing any queries on the com.company.model.Request object types in the BDM, a verification has to be run. In this case we indicate that the verification is a groovy script, *SubjectTeacherPermissionRule.groovy*, which we will now create. For more information dynamic security and how it works with Bonita, see [Rest API authorization](rest-api-authorization.md) ...
````
For more information **about** dynamic security...

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -373,7 +373,7 @@ For the details on how to easily create and map those profiles see the [Profile
 **3. Create a process and generate some request instances**
 
-You will need some instances of the request object, as well as some students. To create them, follow the steps discribed in the section  [II - Invoice](#bdmFilling).
+You will need some instances of the request object, as well as some students. To create them, follow the steps discribed in the section  [II - Invoice](#bdmFilling). For convinience, we assume the only two subjects are Mathematics and Physics.
```
conv**e**nience

educhastenier commented on md/bdm-access-control.md:
```diff
@@ -314,3 +315,163 @@ Deploy your access control file.
+| content | STRING |
+| student | Student |
+
+Subject and Student are mandatory fields. Student is in an eager, aggregated relationship.
```
I'm not sure `eager` is a term that is used in the User Interface... can you check?

danila-m replied:
We say "always load related objects" ... kind of awkward to use those terms.

NathalieC replied:
you can use "Student is an eager (related objects are always loaded), aggregated relationship"

danila-m committed:
fix typos

NathalieC commented on md/bdm-access-control.md:
```diff
@@ -314,3 +315,163 @@ Deploy your access control file.
+
+### III - Instances protection
+
+Instance protection is not supported as a part of the BDM Access Control feature. However it is still possible to solve most use cases using the [rest-api authorizations](rest-api-authorization.md).
```
No such negative sentence should be found in the documentation.
"To grant access to specific instances of business objects, use ...."

NathalieC commented on md/bdm-access-control.md:
```diff
@@ -314,3 +315,163 @@ Deploy your access control file.
+Following is a step by step guide on how to realize one such (simple) use case:
+
+::: info
+The method shown in the example below is not a new 7.7 feature, it is present in the Community version since 7.0 ( although the location of the files and the push/pull procedure changed in 7.6.0).
```
The example below accounts for a specific way to use a method introduced in Bonita 7.0, and updated in Bonita 7.6.

NathalieC commented on md/bdm-access-control.md:
```diff
@@ -314,3 +315,163 @@ Deploy your access control file.
+
+::: info
+The method shown in the example below is not a new 7.7 feature, it is present in the Community version since 7.0 ( although the location of the files and the push/pull procedure changed in 7.6.0).
+It allows to protect the bdm instances by limiting the access to the BDM query requests that retrieve the object instances rather than to the instances themselves.
```
Does it allow to protect or to grant access?
If this is a white list mechanism, then we should make clear this grants access.
I know, the difference is tiny, but this sets users' mind about the white list way to use the feature.

danila-m replied:
"It allows to protect the bdm instances" -> that's what the user tries to do.

As for the way it works, it's literally one script that returns true or false for a query. Depending on what you put in the script, you get either a black or white list.

NathalieC commented on md/bdm-access-control.md:
```diff
@@ -314,3 +315,163 @@ Deploy your access control file.
+
+In the UI Designer, create an application page (*reviewRequests*):
+
+- Create 2 new variables
```
We usually number the steps with 1. 2. 3., ...

danila-m replied:
The "big" steps are numbered. Here we describe how to create a page in the UIDesigner. Just like for the rest of the page, those kind of steps are not numbered.

danila-m committed:
nath feedback integration 1

NathalieC committed:
doc(bdm ac): revamped the page.

add numbering
create consistency between examples
changed some details: add experienced sales (II), teachers is mapped on a role (III)

NathalieC committed:
Merge branch '7.7' into instance_security_bdm
